File: app/src/main/java/com/jimzrt/umsmounter/Activities/MainActivity.java
package com.jimzrt.umsmounter.Activities;

import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.crashlytics.android.Crashlytics;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.jimzrt.umsmounter.Fragments.DownloadFragment;
import com.jimzrt.umsmounter.Fragments.ImageCreationFragment;
import com.jimzrt.umsmounter.Fragments.MainFragment;
import com.jimzrt.umsmounter.Model.DownloadItem;
import com.jimzrt.umsmounter.Model.ImageItem;
import com.jimzrt.umsmounter.R;
import com.jimzrt.umsmounter.Tasks.BaseTask;
import com.jimzrt.umsmounter.Tasks.CheckFolderTask;
import com.jimzrt.umsmounter.Tasks.CheckMassStorageTask;
import com.jimzrt.umsmounter.Tasks.CheckRootTask;
import com.jimzrt.umsmounter.Tasks.SetPathsTask;
import com.jimzrt.umsmounter.Utils.BackgroundTask;
import com.jimzrt.umsmounter.Utils.Helper;

import java.util.UUID;

public class MainActivity extends AppCompatActivity implements ImageCreationFragment.OnImageCreationListener, DownloadFragment.OnImageDownloadListener {


    public static final String ROOTDIR = "/UMSMounter";

    public static String ROOTPATH = "/data/media/0/UMSMounter";
    public static String USERPATH = "/sdcard/UMSMounter";


    // private static final int READ_REQUEST_CODE = 42;
    // TextView logView = null;
    // Spinner usbMode = null;

    public static String android_id = UUID.randomUUID().toString().substring(31);


    MainFragment mainFragment;
    ImageCreationFragment createImageFragment;
    DownloadFragment downloadFragment;

    Fragment currentFragment;

    Toolbar toolbar;
    private DrawerLayout mDrawerLayout;
    private NavigationView navigationView;

    //private ActionBarDrawerToggle mDrawerToggle;


    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {


        //  Helper.verifyStoragePermissions(this);

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mDrawerLayout = findViewById(R.id.drawer_layout);
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        //getSupportActionBar().setLogo(R.drawable.ic_logo);
        //  getSupportActionBar().setDisplayUseLogoEnabled(true);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);

            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_menu);
        }


        // Check that the activity is using the layout version with
        // the fragment_container FrameLayout
        if (findViewById(R.id.fragment_container) != null) {

            // However, if we're being restored from a previous state,
            // then we don't need to do anything and should return or else
            // we could end up with overlapping fragments.
            if (savedInstanceState != null) {
                return;
            }

            // Create a new Fragment to be placed in the activity layout
            mainFragment = new MainFragment();
            createImageFragment = new ImageCreationFragment();
            downloadFragment = new DownloadFragment();

            // In case this activity was started with special instructions from an
            // Intent, pass the Intent's extras to the fragment as arguments
            // firstFragment.setArguments(getIntent().getExtras());

            SharedPreferences sharedPref = getSharedPreferences(null, Context.MODE_PRIVATE);
            boolean firstRun = sharedPref.getBoolean("firstRun", true);
            if (firstRun) {
                checkAll();
            } else {
                // Add the fragment to the 'fragment_container' FrameLayout
                getSupportFragmentManager().beginTransaction()
                        .add(R.id.fragment_container, mainFragment).commit();
            }



        }

        currentFragment = mainFragment;

        navigationView = findViewById(R.id.nav_view);
        navigationView.getMenu().getItem(0).setChecked(true);
        navigationView.setNavigationItemSelectedListener(
                menuItem -> {
                    // set item as selected to persist highlight
                    menuItem.setChecked(true);
                    // Log.i("lala", ""+currentFragmentId);
                    // close drawer when item is tapped
                    mDrawerLayout.closeDrawers();

                    switch (menuItem.getItemId()) {
                        case R.id.nav_home:
                            if (currentFragment != mainFragment) {
                                showMain();

                            } else {
                                Toast.makeText(this, "Already on this mothafucka", Toast.LENGTH_LONG).show();
                            }
                            break;
                        case R.id.nav_create_image:
                            if (currentFragment != createImageFragment) {
                                showCreateImage();
                            } else {
                                Toast.makeText(this, "Already on this mothafucka", Toast.LENGTH_LONG).show();
                            }
                            break;
                        case R.id.nav_download_image:
                            if (currentFragment != downloadFragment) {
                                showDownloadImage();
                            } else {
                                Toast.makeText(this, "Already on this mothafucka", Toast.LENGTH_LONG).show();
                            }
                            break;

                    }
                    // Add code here to update the UI based on the item selected
                    // For example, swap UI fragments here

                    return true;
                });


        //  }
        getSupportFragmentManager().addOnBackStackChangedListener(
                () -> {
                    if (mainFragment.isVisible()) {
                        navigationView.setCheckedItem(R.id.nav_home);
                        currentFragment = mainFragment;
                    } else if (createImageFragment.isAdded()) {
                        navigationView.setCheckedItem(R.id.nav_create_image);
                        currentFragment = createImageFragment;
                    } else if (downloadFragment.isAdded()) {
                        navigationView.setCheckedItem(R.id.nav_download_image);
                        currentFragment = downloadFragment;
                    } else {
                        navigationView.setCheckedItem(-1);
                        currentFragment = downloadFragment;
                    }

                });


        Helper.trustAllHosts();
    }

    private void showDownloadImage() {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.setCustomAnimations(
                R.anim.card_flip_right_in,
                R.anim.card_flip_right_out,
                R.anim.card_flip_left_in,
                R.anim.card_flip_left_out);

        if (currentFragment == mainFragment) {
            transaction.hide(mainFragment);
        } else {
            transaction.remove(currentFragment);
        }
        transaction.add(R.id.fragment_container, downloadFragment);


        transaction.addToBackStack(null);
        transaction.commit();
        currentFragment = downloadFragment;
        navigationView.setCheckedItem(R.id.nav_download_image);
    }

    @Override
    public void onStart() {
        super.onStart();


    }

    private void showCreateImage() {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.setCustomAnimations(
                R.anim.card_flip_right_in,
                R.anim.card_flip_right_out,
                R.anim.card_flip_left_in,
                R.anim.card_flip_left_out);
        if (currentFragment == mainFragment) {
            transaction.hide(mainFragment);
        } else {
            transaction.remove(currentFragment);
        }
        transaction.add(R.id.fragment_container, createImageFragment);
        transaction.addToBackStack(null);
        transaction.commit();
        currentFragment = createImageFragment;
        navigationView.setCheckedItem(R.id.nav_create_image);
    }

    private void showMain() {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.setCustomAnimations(
                R.anim.card_flip_left_in,
                R.anim.card_flip_left_out,
                R.anim.card_flip_right_in,
                R.anim.card_flip_right_out);
        transaction.remove(currentFragment);
        transaction.show(mainFragment);
        //transaction.addToBackStack(null);
        transaction.commit();
        currentFragment = mainFragment;
        navigationView.setCheckedItem(R.id.nav_home);
        getSupportFragmentManager().popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();

        switch (id) {
            case R.id.action_revert:
                mainFragment.unmount("mtp,adb");
                return true;
            case R.id.action_check_dependencies:
                checkAll();
                return true;
            case android.R.id.home:
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;

        }


        return super.onOptionsItemSelected(item);
    }


    void checkAll() {


        SharedPreferences sharedPref = getSharedPreferences(null, Context.MODE_PRIVATE);
        (new BackgroundTask(this).setDelegate((successful, output) -> {
            if (successful) {


                SharedPreferences.Editor editor = sharedPref.edit();
                editor.putBoolean("firstRun", false);
                editor.apply();

                if (!mainFragment.isAdded()) {
                    // Add the fragment to the 'fragment_container' FrameLayout
                    getSupportFragmentManager().beginTransaction()
                            .add(R.id.fragment_container, mainFragment).commit();
                }


            } else {
                SharedPreferences.Editor editor = sharedPref.edit();
                editor.clear().apply();
                AlertDialog.Builder builder = new AlertDialog.Builder(this);

                builder.setMessage(output + "\n\nAfter clicking Ok the app will crash to send a report so the issue can be fixed!\nYour id: " + android_id)
                        .setTitle("Error!");
                builder.setPositiveButton("Ok", (dialog, which) -> {
                    Crashlytics.setUserIdentifier(android_id);
                    Crashlytics.log(1, "Error", output);
                    Crashlytics.getInstance().crash();
                });

                builder.setNegativeButton("Close App", (dialog, which) -> {
                    finish();

                });

                AlertDialog dialog = builder.create();
                dialog.show();
            }
        })).setTasks(new BaseTask[]{new CheckRootTask(), new CheckFolderTask(), new SetPathsTask(), new CheckMassStorageTask()}).execute();
    }

    @Override
    public void OnImageCreation(String imageItemName) {
        showMain();
        ImageItem imageItem = new ImageItem(imageItemName, ROOTPATH + "/" + imageItemName, USERPATH + "/" + imageItemName, Helper.humanReadableByteCount(0));
        mainFragment.createImage(imageItem);

    }

    @Override
    public void OnImageListClick(DownloadItem downloadItem) {

        Gson gson = new GsonBuilder().create();
        String downloadItemString = gson.toJson(downloadItem);

        Intent intent = new Intent(this, LinuxImageActivity.class);


        intent.putExtra("downloadItem", downloadItemString);
        startActivityForResult(intent, 0);


    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        // Check which request we're responding to
        if (requestCode == 0) {
            // Make sure the request was successful
            if (resultCode == RESULT_OK) {
//
                String name = data.getStringExtra("name");
                String url = data.getStringExtra("url");
                ImageItem imageItem = new ImageItem(name, ROOTPATH + "/" + name, USERPATH + "/" + name, Helper.humanReadableByteCount(0));
                imageItem.setUrl(url);
                showMain();
                mainFragment.addImage(imageItem);

            }
        }
    }
}

File: app/src/main/java/com/jimzrt/umsmounter/activities/MainActivity.kt
package com.jimzrt.umsmounter.activities

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import com.google.android.material.navigation.NavigationView
import com.google.gson.GsonBuilder
import com.jimzrt.umsmounter.BuildConfig
import com.jimzrt.umsmounter.R
import com.jimzrt.umsmounter.fragments.CreditsFragment
import com.jimzrt.umsmounter.fragments.DownloadFragment
import com.jimzrt.umsmounter.fragments.DownloadFragment.OnImageDownloadListener
import com.jimzrt.umsmounter.fragments.ImageCreationFragment
import com.jimzrt.umsmounter.fragments.ImageCreationFragment.OnImageCreationListener
import com.jimzrt.umsmounter.fragments.MainFragment
import com.jimzrt.umsmounter.model.DownloadItem
import com.jimzrt.umsmounter.model.ImageItem
import com.jimzrt.umsmounter.tasks.*
import com.jimzrt.umsmounter.utils.BackgroundTask
import com.jimzrt.umsmounter.utils.Helper

class MainActivity : AppCompatActivity(), OnImageCreationListener, OnImageDownloadListener {
    private var mainFragment: MainFragment? = null
    private var createImageFragment: ImageCreationFragment? = null
    private var downloadFragment: DownloadFragment? = null
    private var creditsFragment: CreditsFragment? = null
    private var currentFragment: Fragment? = null
    private var mDrawerLayout: DrawerLayout? = null
    private var navigationView: NavigationView? = null
    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        mDrawerLayout = findViewById(R.id.drawer_layout)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        if (supportActionBar != null) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            supportActionBar!!.setHomeButtonEnabled(true)
            supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_menu)
        }
        if (findViewById<View?>(R.id.fragment_container) != null) {
            if (savedInstanceState != null) {
                return
            }

            // Create a new Fragment to be placed in the activity layout
            mainFragment = MainFragment()
            createImageFragment = ImageCreationFragment()
            downloadFragment = DownloadFragment()
            creditsFragment = CreditsFragment()
            val sharedPref = getSharedPreferences(null, Context.MODE_PRIVATE)
            val firstRun = sharedPref.getBoolean("firstRun", true)
            val version = sharedPref.getString("version", "")
            USERPATH = sharedPref.getString("userpath", "")
            ROOTPATH = sharedPref.getString("rootpath", "")
            if (firstRun || BuildConfig.VERSION_NAME != version) {
                checkAll()
            } else {
                // Add the fragment to the 'fragment_container' FrameLayout
                supportFragmentManager.beginTransaction()
                        .add(R.id.fragment_container, mainFragment!!).commit()
            }
        }
        currentFragment = mainFragment
        navigationView = findViewById(R.id.nav_view)
        navigationView?.menu?.getItem(0)?.isChecked = true
        navigationView?.setNavigationItemSelectedListener { menuItem: MenuItem ->
            // set item as selected to persist highlight
            // menuItem.setChecked(true);

            //   mDrawerLayout.closeDrawers();
            val delay = 150
            when (menuItem.itemId) {
                R.id.nav_home -> if (currentFragment !== mainFragment) {
                    Handler().postDelayed({ showMain() }, delay.toLong())
                }
                R.id.nav_create_image -> if (currentFragment !== createImageFragment) {
                    Handler().postDelayed({ showCreateImage() }, delay.toLong())
                }
                R.id.nav_download_image -> if (currentFragment !== downloadFragment) {
                    Handler().postDelayed({ showDownloadImage() }, delay.toLong())
                }
                R.id.nav_credits -> if (currentFragment !== createImageFragment) {
                    Handler().postDelayed({ showCredits() }, delay.toLong())
                }
            }
            // Add code here to update the UI based on the item selected
            // For example, swap UI fragments here
            mDrawerLayout?.closeDrawer(GravityCompat.START)
            true
        }


        //  }
        supportFragmentManager.addOnBackStackChangedListener {
            when {
                mainFragment!!.isVisible -> {
                    navigationView?.setCheckedItem(R.id.nav_home)
                    currentFragment = mainFragment
                }
                createImageFragment!!.isAdded -> {
                    navigationView?.setCheckedItem(R.id.nav_create_image)
                    currentFragment = createImageFragment
                }
                downloadFragment!!.isAdded -> {
                    navigationView?.setCheckedItem(R.id.nav_download_image)
                    currentFragment = downloadFragment
                }
                creditsFragment!!.isAdded -> {
                    navigationView?.setCheckedItem(R.id.nav_credits)
                    currentFragment = creditsFragment
                }
            }
        }


        //Helper.trustAllHosts();
    }

    private fun showDownloadImage() {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.setCustomAnimations(android.R.anim.fade_in,
                android.R.anim.fade_out)
        if (currentFragment === mainFragment) {
            transaction.hide(mainFragment!!)
        } else {
            transaction.remove(currentFragment!!)
        }
        transaction.add(R.id.fragment_container, downloadFragment!!)
        transaction.addToBackStack(null)
        transaction.commit()
        currentFragment = downloadFragment
        navigationView!!.setCheckedItem(R.id.nav_download_image)
    }

    private fun showCredits() {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.setCustomAnimations(android.R.anim.fade_in,
                android.R.anim.fade_out)
        if (currentFragment === mainFragment) {
            transaction.hide(mainFragment!!)
        } else {
            transaction.remove(currentFragment!!)
        }
        transaction.add(R.id.fragment_container, creditsFragment!!)
        transaction.addToBackStack(null)
        transaction.commit()
        currentFragment = creditsFragment
        navigationView!!.setCheckedItem(R.id.nav_credits)
    }

    public override fun onStart() {
        super.onStart()
    }

    private fun showCreateImage() {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.setCustomAnimations(android.R.anim.fade_in,
                android.R.anim.fade_out)
        if (currentFragment === mainFragment) {
            transaction.hide(mainFragment!!)
        } else {
            transaction.remove(currentFragment!!)
        }
        transaction.add(R.id.fragment_container, createImageFragment!!)
        transaction.addToBackStack(null)
        transaction.commit()
        currentFragment = createImageFragment
        navigationView!!.setCheckedItem(R.id.nav_create_image)
    }

    private fun showMain() {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.setCustomAnimations(android.R.anim.fade_in,
                android.R.anim.fade_out)
        transaction.remove(currentFragment!!)
        transaction.show(mainFragment!!)
        //transaction.addToBackStack(null);
        transaction.commit()
        currentFragment = mainFragment
        navigationView!!.setCheckedItem(R.id.nav_home)
        supportFragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_revert -> {
                mainFragment!!.unmount("mtp,adb")
                return true
            }
            R.id.action_check_dependencies -> {
                checkAll()
                return true
            }
            android.R.id.home -> {
                mDrawerLayout!!.openDrawer(GravityCompat.START)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun checkAll() {
        val sharedPref = getSharedPreferences(null, Context.MODE_PRIVATE)
        val context = this
        BackgroundTask(this).setDelegate(object : BackgroundTask.AsyncResponse {
            override fun processFinish(successful: Boolean?, output: String?) {
                if (successful!!) {
                    val editor = sharedPref.edit()
                    editor.putBoolean("firstRun", false)
                    editor.putString("version", BuildConfig.VERSION_NAME)
                    editor.apply()
                } else {
                    val editor = sharedPref.edit()
                    editor.clear().apply()
                    val builder = AlertDialog.Builder(context)
                    builder.setMessage(output)
                            .setTitle("Error!")
                    builder.setPositiveButton("Ok", null)
                    val dialog = builder.create()
                    dialog.show()
                }
                if (!mainFragment!!.isAdded) {
                    // Add the fragment to the 'fragment_container' FrameLayout
                    supportFragmentManager.beginTransaction()
                            .add(R.id.fragment_container, mainFragment!!).commit()
                }
            }
        }
        ).setTasks(arrayOf(CheckRootTask(), CheckPermissionTask(), SetPathsTask(), CheckFolderTask(), CheckMassStorageTask())).execute()
    }

    override fun OnImageCreation(imageItem: String?) {
        showMain()
        val imageItemObj = ImageItem(imageItem!!, "$ROOTPATH/$imageItem", "$USERPATH/$imageItem", Helper.humanReadableByteCount(0))
        mainFragment!!.createImage(imageItemObj)
    }

    override fun OnImageListClick(downloadItem: DownloadItem?) {
        val gson = GsonBuilder().create()
        val downloadItemString = gson.toJson(downloadItem)
        val intent = Intent(this, LinuxImageActivity::class.java)
        intent.putExtra("downloadItem", downloadItemString)
        startActivityForResult(intent, 0)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        // Check which request we're responding to
        if (requestCode == 0) {
            // Make sure the request was successful
            if (resultCode == Activity.RESULT_OK) {
//
                val name = data!!.getStringExtra("name")
                val url = data.getStringExtra("url")
                val imageItem = ImageItem(name!!, "$ROOTPATH/$name", "$USERPATH/$name", Helper.humanReadableByteCount(0))
                imageItem.url = url
                showMain()
                mainFragment!!.addImage(imageItem)
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == WRITE_EXTERNAL_STORAGE_PERM) {
            val sharedPref = getSharedPreferences(null, Context.MODE_PRIVATE)
            val editor = sharedPref.edit()
            editor.putBoolean("hasPermission", true)
            editor.apply()
            Toast.makeText(this, "granteddd!!!", Toast.LENGTH_LONG).show()
        } else {
            Toast.makeText(this, "dont know this shit", Toast.LENGTH_LONG).show()
        }
    }

    companion object {
        const val ROOTDIR = "/UMSMounter"
        const val CACHEDIR = "/cache"

        @JvmField
        var ROOTPATH: String? = null

        @JvmField
        var USERPATH: String? = null
        const val WRITE_EXTERNAL_STORAGE_PERM = 1337
    }
}